### ISSUE TITLE
exec plugin buffer overflow under certain conditions

### ISSUE THREAD

**Issue by kwiatrox** [collectd/collectd#3322]
*   Version of collectd: 5.9.2, also master
*   Operating system / distribution: Ubuntu 18.04
*   Kernel version (if applicable): N/A

## Expected behavior

Exec plugin can work with file descriptors bigger than 1024. This situation happen when limit for collectd is increased by `ulimit -n` and other plugin used all the smaller FDs (e.g. intel_pmu plugin).

## Actual behavior

Buffer overflow -> core dump
```
[2019-10-24 14:22:08] [info] Initialization complete, entering read-loop.
[2019-10-24 14:22:08] [debug] Created new plugin context.
[2019-10-24 14:22:08] [debug] intel_pmu: pmu_read:416
[2019-10-24 14:22:08] [debug] Created new plugin context.
[2019-10-24 14:22:08] [debug] plugin_read_thread: read-function of the `exec' plugin took 0.000256 seconds.
[2019-10-24 14:22:08] [debug] plugin_read_thread: Effective interval of the `exec' plugin is 10.000 seconds.
[2019-10-24 14:22:08] [debug] plugin_read_thread: Next read of the `exec' plugin at 1571919738.133.
[2019-10-24 14:22:08] [debug] Created new plugin context.
*** buffer overflow detected ***: /opt/collectd/sbin/collectd terminated
Aborted (core dumped)
```

Parts of core dump:
```
#4  0x00007f1ea8501d21 in __GI___fortify_fail (msg=msg@entry=0x7f1ea8583905 "buffer overflow detected") at fortify_fail.c:44
#5  0x00007f1ea84ffa10 in __GI___chk_fail () at chk_fail.c:28
#6  0x00007f1ea8501c0a in __fdelt_chk (d=<optimized out>) at fdelt_chk.c:25
#7  0x00007f1ea7fbdc25 in exec_read_one (arg=0x5564d9a89120) at src/exec.c:602
        __d = <optimized out>
        pl = 0x5564d9a89120
        fd = 2309
        fd_err = 2311
        __PRETTY_FUNCTION__ = "exec_read_one"
(...)
```

## Steps to reproduce

*   change ulimit
`ulimit -n 4096`
*   setup intel_pmu plugin to monitor all cores and events, assuming that
    "num_of_cores" * "num_of_events" >= 1024
*   enable exec plugin with at least one exec to run
*   start collectd

## Root cause

Use of select() and FD_*() macros in exec_read_one() function. Core dump is caused by line:
`FD_SET(fd, &fdset);`

From select() documentation:
```
select() can monitor only file descriptors numbers that are less than FD_SETSIZE;
poll(2) does not have this limitation.
BUGS
 POSIX allows an implementation to define an upper limit, advertised via the constant FD_SETSIZE,
on the range of file descriptors that can be specified in a file descriptor set.
 The Linux kernel imposes no fixed limit, but the glibc implementation makes fd_set a fixed-size
type, with FD_SETSIZE defined as 1024, and the FD_*() macros operating according to that limit.
To monitor file descriptors greater than 1023, use poll(2) instead.
```

The solution would be to replace `select()` with `poll()`.

**Comment by kwiatrox** [comment]
Fixed with #3363 - closing.

**Comment by rpv-tomsk** [comment]
Do not confuse “fixed” and “should be fixed”.
